Add and option to have a button that opens an emoji keyboard to the Shacknews post form

TroZ opened the pull request:
Adds and option to have a button that opens an emoji keyboard to the Shacknews post form which makes including emoji and other special characters easy.
Modifies post preview to support previewing of emoji characters.
Includes options for enabling the keyboard, and having the button before or after the post button (technically as the first button or the last button, so if the preview button is enabled and on the same side, the preview button will always be next to the post button).
Emoji keyboard has a 'Unicode 6' view / order and a 'Android' view / order for the tabs / characters. I'd like to add an 'iOS' order but currently don't have a reference for the emoji keyboard in iOS that I know is accurate.

TroZ commented:
I believe I've made the requested changed.
I'm not entirely sure if you will get a notice of this, or if I have to submit a new pull request.

electroly commented:
What I originally envisioned for emoji support is:

1.) Rig the post button to encode the entities before posting, so that emoji can be typed literally into the post box instead of the user seeing the entities.

2.) Add a platform-specific message near the post box telling Windows users to press Win+Semicolon and telling macOS users to press Ctrl+Command+Space. This will show the system emoji keyboard.

I still feel like this is the better approach. I'm not sure I like the idea of piling on more third party libraries and shipping binary fonts to do what the OS already does. Is there a compelling reason to use the third party emoji picker instead of the OS one?

TroZ commented:
For point 1, you are correct, that would be a much better approach for emoji support.  I had written this 'emoji keyboard' a number of years ago, and originally was just thinking of the easiest way to have it work with Shacknews.  It will probably take me a few days to get the code working, but encoding the entities just before they are posted would be a much better approach.
For point 2, I was not aware of Win+Semicolon. I can add a message to that effect to the post form.  However, I still think a javascript emoji window like I implemented is useful for a few reasons: 1. The windows emoji entry does not allow selection of families with mixed skin tones like my emoji window does. 2. My window includes interesting characters that are technically not emoji, but some people would find useful, that are normally difficult to type (foreign currency characters, for example) . 3. People may be more familiar with Android's or iOS's ordering of emoji characters, which the custom emoji keyboard can replicate (it currently does replicate Android's, and if I can find an accurate source iOS's ordering of emoji tabs / characters, I can add in an iOS mode as well).

I propose that I should implement the parsing of emoji characters just before posting as the default behavior for this addition, with a message for how to open the OS's emoji keyboard.  Then as an option, the custom emoji keyboard can be turned on (with option for button location). I'll attempt to re-work the custom keyboard to remove reliance on the fonts, which will remove all third party elements of this change. This will probably require removing the hieroglyphics, and a few other sections of characters (as well as the few Unicode 12 characters that I already added, as they aren't in the window's fonts yet), but will hopefully still leave most of the interesting characters rendering with the user's already installed fonts.

boarder2 commented:
Windows emoji keyboard does support skin tones on the people tab. In general I think it's best to stick with the platform native option when available.

TroZ commented:
While the Windows emoji picker does support skin tones (I didn't say it didn't), it does support mixed skin tones for families or love/kiss characters, such as:  👩🏼‍👨🏿‍👧‍👦🏽  or 👨🏾‍❤️‍💋‍👩🏼.  The Window's picker doesn't support skin tones at all for these characters, and doesn't support different skin tones per person, nor do does support different orders of people (The window picker always has the man first in the family emoji if there is both a man and a woman, and the woman first for the love/kiss emoji if there is both a woman and man). As these are both made up of multiple emojis joined with zero width joiner characters, Unicode supports any ordering and having different skin tones per person (it should technically also support different hair styles per person as well, but apparently the windows font renderer doesn't support hair style when the character isn't just a full face).
I will make changes to support the OS's emoji picker.  However, I think providing an optional picker that allows a larger selection of characters is also a good thing, and can be done while also removing the third party fonts that I previously used.

electroly commented:
Removing the fonts seems logical to me for Chrome Shack -- even if the poster has the fonts, the other Shackers reading the post won't. I'd prefer the picker to contain "chatty safe" symbols that can be posted with a reasonable expectation that everyone (besides perhaps Lamp users) will be able to see it.

Including a third party library isn't the end of the world, we just have to track the dependency and the source version for the Firefox store submissions. They require us to document our dependencies and they actually inspect the code to check.

My general philosophy is that if someone wants to do a bunch of work to add something to Chrome Shack, the default answer is to try to accept the addition unless the addition is clearly problematic. I don't think the addition of a JavaScript emoji picker meets that standard. Seems like there is a reasonable argument for its use. I use Slack's emoji picker even though Windows has one.

TroZ commented:
I have removed external fonts, and changed to code to work differently.
Now the system emoji picker can be used to put emoji characters directly in to the post form (with text below the post form added for how to open the system emoji picker). If enabled, my script disables the normal post process, and adds it's own handler to the post button that converts and characters outside the basic multilingual plane (any characters represented with a surrogate pair) into an html entity, before submitting the post form. It also modifies the post preview to not remove surrogate pairs, so that the emoji are previewed properly.
Optionally, my custom emoji picker can be enabled, and it will add a button either before or after the post button to open it. I had to remove a few groups of characters that the fonts allowed to be displayed, but I'm actually surprised that many of the characters render without the fonts being used (hieroglyphics and cuneiform I was sure would need to be removed).

electroly commented:
I promise I haven't forgotten this. I haven't gotten a chance to review it. We will have to ship a new Chrome Shack version for lol changes coming soon, I'll ship this feature in that release.

electroly commented:
I gave this a quick test and I couldn't seem to get the button to appear. I enabled the feature in the options and reloaded the chatty, but there was no button in the post box area. 

I did see the "Press Win + : ..." text, so at least something is happening. Two notes about that: on Windows it's semicolon (;) and not colon (:). Also, I was testing on Linux and it showed me the Windows shortcut. I wasn't sure if it was detecting Mac vs. Windows and not including Linux, or if there is no detection. Linux is pretty niche, probably safe to just show nothing if it's not Windows and not Mac.

Any idea about the missing button? I fiddled with the positioning toggle but it didn't make the button appear.

TroZ commented:
I believe I have fixed the issue.
Apparently I was using .selected when I should have been using .checked for
the emoji keyboard button.  How it was seeming working for me, I'm not
sure, other than perhaps I enabled it and never tried disabling it again.
Additionally, I now disable the position radio buttons if the emoji
keyboard is not enabled.
I've updated my branch with that fix, and additionally should only be
showing the system emoji keyboard message for Mac or Windows.

On Thu, May 30, 2019 at 10:50 AM Brian Risinger <brian.risinger@gmail.com>
wrote:

> I'll take a closer look tonight, but I could not reproduce the button not
> showing up if the emoji processing and the emoji keyboard check boxes are
> enabled.
>
> The message appears if emoji processing is enabled, but for the button to
> appear, the emoji keyboard checkbox must also be checked.  Apparently I
> enable the position radio buttons even if the emoji keyboard is not
> enabled.  I did just test the different combinations of the emoji keyboard
> button along with different position of the preview button or having the
> preview button disabled, and the emoji keyboard button always appeared.
> I'll take another look after work tonight; perhaps some other option
> modifies the html of the button line that I'm not aware of, and that
> interferes with the button being added.
>
>
> Is there a keyboard combination for Linux that will show an emoji
> keyboard?  I think the way I coded it, if it doesn't see 'Mac' in the user
> agent it assumes Windows, but I can change that to it will only show the
> Windows shortcut if it sees 'Win' in the user agent.
> And actually, for me, the shortcut for Windows isn't working since the May
> update for Windows came out. I'm not sure if they added a switch for it in
> the settings somewhere, it if it somehow broke.  It worked when I initially
> submitted the pull request.
>
> On Thu, May 30, 2019 at 10:23 AM Brian Luft <notifications@github.com>
> wrote:
>
>> I gave this a quick test and I couldn't seem to get the button to appear.
>> I enabled the feature in the options and reloaded the chatty, but there was
>> no button in the post box area.
>>
>> I did see the "Press Win + : ..." text, so at least something is
>> happening. Two notes about that: on Windows it's semicolon (;) and not
>> colon (:). Also, I was testing on Linux and it showed me the Windows
>> shortcut. I wasn't sure if it was detecting Mac vs. Windows and not
>> including Linux, or if there is no detection. Linux is pretty niche,
>> probably safe to just show nothing if it's not Windows and not Mac.
>>
>> Any idea about the missing button? I fiddled with the positioning toggle
>> but it didn't make the button appear.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/latestchatty/chromeshack/pull/125?email_source=notifications&email_token=AAC22IYMDCPREQKWFGXVLRTPX7PPNA5CNFSM4HHKQGXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWSO22Q#issuecomment-497347946>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAC22I56CXCAXHRM6MBH5ETPX7PPNANCNFSM4HHKQGXA>
>> .
>>
>

TroZ commented:
I'll take a closer look tonight, but I could not reproduce the button not
showing up if the emoji processing and the emoji keyboard check boxes are
enabled.

The message appears if emoji processing is enabled, but for the button to
appear, the emoji keyboard checkbox must also be checked.  Apparently I
enable the position radio buttons even if the emoji keyboard is not
enabled.  I did just test the different combinations of the emoji keyboard
button along with different position of the preview button or having the
preview button disabled, and the emoji keyboard button always appeared.
I'll take another look after work tonight; perhaps some other option
modifies the html of the button line that I'm not aware of, and that
interferes with the button being added.


Is there a keyboard combination for Linux that will show an emoji
keyboard?  I think the way I coded it, if it doesn't see 'Mac' in the user
agent it assumes Windows, but I can change that to it will only show the
Windows shortcut if it sees 'Win' in the user agent.
And actually, for me, the shortcut for Windows isn't working since the May
update for Windows came out. I'm not sure if they added a switch for it in
the settings somewhere, it if it somehow broke.  It worked when I initially
submitted the pull request.

On Thu, May 30, 2019 at 10:23 AM Brian Luft <notifications@github.com>
wrote:

> I gave this a quick test and I couldn't seem to get the button to appear.
> I enabled the feature in the options and reloaded the chatty, but there was
> no button in the post box area.
>
> I did see the "Press Win + : ..." text, so at least something is
> happening. Two notes about that: on Windows it's semicolon (;) and not
> colon (:). Also, I was testing on Linux and it showed me the Windows
> shortcut. I wasn't sure if it was detecting Mac vs. Windows and not
> including Linux, or if there is no detection. Linux is pretty niche,
> probably safe to just show nothing if it's not Windows and not Mac.
>
> Any idea about the missing button? I fiddled with the positioning toggle
> but it didn't make the button appear.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/latestchatty/chromeshack/pull/125?email_source=notifications&email_token=AAC22IYMDCPREQKWFGXVLRTPX7PPNA5CNFSM4HHKQGXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWSO22Q#issuecomment-497347946>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAC22I56CXCAXHRM6MBH5ETPX7PPNANCNFSM4HHKQGXA>
> .
>

WombatFromHell commented:
I went with a significantly leaner approach in commit 6406ec36e21bfb678e2807ec1fa09bc9c945199a.

electroly commented:
Thanks again for this PR, @TroZ. We have credited you in the release notes for this feature.
